File: packages/sdk-js/src/resources/api_keys.ts
import { BaseResource } from './base';
import { AxiosRequestConfig } from 'axios';

// Define interfaces for API Key data structures
interface ApiKey {
  id: string;
  name: string;
  key?: string; // Key is only returned on create/regenerate
  type: 'api' | 'admin' | 'webhook';
  createdAt: string; // ISO Date string
  lastUsed?: string | null; // ISO Date string
  expiresAt?: string | null; // ISO Date string
  revoked: boolean;
  scopes?: string[]; // Added based on spec
}

interface CreateApiKeyPayload {
  name: string; // 3-64 chars
  type: 'api' | 'admin' | 'webhook';
  expiresAt?: string | null; // ISO Date string
  scopes?: string[];
}

interface UpdateApiKeyPayload {
  name?: string; // 3-64 chars
  expiresAt?: string | null; // ISO Date string or null to remove expiration
  revoked?: boolean;
  scopes?: string[];
}

interface RegenerateApiKeyPayload {
    retainName?: boolean;
    expiresAt?: string | null; // ISO Date string
}

export class ApiKeysResource extends BaseResource {
  protected readonly RESOURCE_PATH = "api-keys"; // Note hyphen

  async list(config?: AxiosRequestConfig): Promise<Omit<ApiKey, 'key'>[]> {
    // GET /api/api-keys - Key is typically omitted in list view
    return this._list<Omit<ApiKey, 'key'>[]>(undefined, config);
  }

  async create(payload: CreateApiKeyPayload, config?: AxiosRequestConfig): Promise<ApiKey> {
    // POST /api/api-keys - Returns the full key on creation
     const cleanedPayload = Object.entries(payload)
        .filter(([_, value]) => value !== undefined) // Keep nulls if intended
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {});
    return this._create<ApiKey>(cleanedPayload, undefined, config);
  }

  async get(keyId: string, config?: AxiosRequestConfig): Promise<Omit<ApiKey, 'key'>> {
    // GET /api/api-keys/{id} - Key is typically omitted in get view
    return this._get<Omit<ApiKey, 'key'>>(keyId, undefined, config);
  }

  async update(keyId: string, payload: UpdateApiKeyPayload, config?: AxiosRequestConfig): Promise<Omit<ApiKey, 'key'>> {
    // PUT /api/api-keys/{id}
     const cleanedPayload = Object.entries(payload)
        .filter(([_, value]) => value !== undefined) // Keep nulls if intended
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {});
    return this._update<Omit<ApiKey, 'key'>>(keyId, cleanedPayload, undefined, config);
  }

  async delete(keyId: string, config?: AxiosRequestConfig): Promise<null> {
    // DELETE /api/api-keys/{id} - Spec indicates 204 No Content
    return this._delete<null>(keyId, undefined, config);
  }

  async regenerate(keyId: string, payload?: RegenerateApiKeyPayload, config?: AxiosRequestConfig): Promise<ApiKey> {
    // POST /api/api-keys/{id}/regenerate - Returns the full new key
    const action = "regenerate";
    const cleanedPayload = Object.entries(payload || {})
        .filter(([_, value]) => value !== undefined) // Keep nulls if intended
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {});
    return this._action<ApiKey>(keyId, action, 'POST', cleanedPayload, undefined, config);
  }

  async revoke(keyId: string, config?: AxiosRequestConfig): Promise<Omit<ApiKey, 'key'>> {
    // POST /api/api-keys/{id}/revoke
    const action = "revoke";
    return this._action<Omit<ApiKey, 'key'>>(keyId, action, 'POST', undefined, undefined, config);
  }
}

File: packages/sdk-python/veltrix_sdk/resources/api_keys.py
# -*- coding: utf-8 -*-

from .base import BaseResource

class ApiKeysResource(BaseResource):
    """Handles operations related to API keys."""
    RESOURCE_PATH = "api-keys"  # Note the hyphenated path from the spec

    def list(self, **kwargs):
        """
        Retrieves all API keys for the authenticated organization.

        Args:
            **kwargs: Additional keyword arguments for the request.

        Returns:
            list: A list of API key objects (key value is usually masked or omitted).
        """
        # Corresponds to GET /api/api-keys
        return self._list(**kwargs)

    def create(self, name, key_type, expires_at=None, scopes=None, **kwargs):
        """
        Creates a new API key.

        Args:
            name (str): The name for the API key (3-64 characters).
            key_type (str): Type of key ('api', 'admin', or 'webhook').
            expires_at (str, optional): ISO 8601 formatted expiration date. Defaults to None (no expiration).
            scopes (list[str], optional): List of permission scopes. Defaults to None.
            **kwargs: Additional keyword arguments for the request.

        Returns:
            dict: The newly created API key object, including the key value.
        """
        # Corresponds to POST /api/api-keys
        data = {
            "name": name,
            "type": key_type,  # Renamed from 'type' to avoid keyword conflict
            "expiresAt": expires_at,
            "scopes": scopes,
        }
        data = {k: v for k, v in data.items() if v is not None}
        return self._create(data=data, **kwargs)

    def get(self, key_id, **kwargs):
        """
        Retrieves a specific API key by its ID (key value is usually masked or omitted).

        Args:
            key_id (str): The ID of the API key to retrieve.
            **kwargs: Additional keyword arguments for the request.

        Returns:
            dict: The API key object.
        """
        # Corresponds to GET /api/api-keys/{id}
        return self._get(resource_id=key_id, **kwargs)

    def update(self, key_id, name=None, expires_at=None, revoked=None, scopes=None, **kwargs):
        """
        Updates an existing API key.

        Args:
            key_id (str): The ID of the API key to update.
            name (str, optional): New name for the API key (3-64 characters).
            expires_at (str or None, optional): New ISO 8601 expiration date, or None to remove expiration.
            revoked (bool, optional): Set the revoked status.
            scopes (list[str], optional): New list of permission scopes.
            **kwargs: Additional keyword arguments for the request.

        Returns:
            dict: The updated API key object.
        """
        # Corresponds to PUT /api/api-keys/{id}
        data = {
            "name": name,
            "expiresAt": expires_at,
            "revoked": revoked,
            "scopes": scopes,
        }
        data = {k: v for k, v in data.items() if v is not None}
        return self._update(resource_id=key_id, data=data, **kwargs)

    def delete(self, key_id, **kwargs):
        """
        Deletes an API key by its ID.

        Args:
            key_id (str): The ID of the API key to delete.
            **kwargs: Additional keyword arguments for the request.

        Returns:
            None: Returns None on successful (204 No Content) deletion.
        """
        # Corresponds to DELETE /api/api-keys/{id}
        return self._delete(resource_id=key_id, **kwargs)

    def regenerate(self, key_id, retain_name=True, expires_at=None, **kwargs):
        """
        Regenerates an API key, invalidating the old key value.

        Args:
            key_id (str): The ID of the API key to regenerate.
            retain_name (bool, optional): Whether to keep the existing name. Defaults to True.
            expires_at (str, optional): New ISO 8601 expiration date. Defaults to None.
            **kwargs: Additional keyword arguments for the request.

        Returns:
            dict: The regenerated API key object, including the new key value.
        """
        # Corresponds to POST /api/api-keys/{id}/regenerate
        action = "regenerate"
        data = {
            "retainName": retain_name,
            "expiresAt": expires_at,
        }
        data = {k: v for k, v in data.items() if v is not None}
        # Use _action helper as it's a POST to a sub-path
        return self._action(resource_id=key_id, action=action, method='POST', data=data, **kwargs)

    def revoke(self, key_id, **kwargs):
        """
        Revokes an API key, making it inactive.

        Args:
            key_id (str): The ID of the API key to revoke.
            **kwargs: Additional keyword arguments for the request.

        Returns:
            dict: The updated (revoked) API key object.
        """
        # Corresponds to POST /api/api-keys/{id}/revoke
        action = "revoke"
        # Use _action helper as it's a POST to a sub-path with no body
        return self._action(resource_id=key_id, action=action, method='POST', **kwargs)
